src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+@doc doc"""
+    ProductManifold{TM<:Tuple, TRanges<:Tuple, TSizes<:Tuple} <: Manifold
+
+Product manifold $M_1 \times M_2 \times \dots \times M_n$ with product geometry.
+`TRanges` and `TSizes` statically define the relationship between representation
+of the product manifold and representations of point, tangent vectors
+and cotangent vectors of respective manifolds.
+
+# Constructor
+
+    ProductManifold(M_1, M_2, ..., M_n)
+
+generates the product manifold $M_1 \times M_2 \times \dots \times M_n$.
+Alternatively, the same manifold can be contructed using the `×` operator:
+`M_1 × M_2 × M_3`.
+"""
+struct ProductManifold{TM<:Tuple} <: Manifold
+    manifolds::TM
+end
+
+ProductManifold(manifolds::Manifold...) = ProductManifold{typeof(manifolds)}(manifolds)
+
+function cross(M1::Manifold, M2::Manifold)
+    return ProductManifold(M1, M2)
+end
+
+function cross(M1::ProductManifold, M2::Manifold)
+    return ProductManifold(M1.manifolds..., M2)
+end
+
+function cross(M1::Manifold, M2::ProductManifold)
+    return ProductManifold(M1, M2.manifolds...)
+end
+
+function cross(M1::ProductManifold, M2::ProductManifold)
+    return ProductManifold(M1.manifolds..., M2.manifolds...)
+end
+
+"""
+    ShapeSpecification(manifolds::Manifold...)
+
+A structure for specifying array size and offset information for linear
+storage of points and tangent vectors on the product manifold of `manifolds`.
+"""
+struct ShapeSpecification{TRanges, TSizes} end
```

Comment:
Would be nice to have documentation for `TRanges` and `TSizes`.

Reply:
Right, I've prepared documentation for it.

Reply:
Awesome, thanks! It's very clear now. It would be convenient if the user didn't need to explicitly create a `ShapeSpecification`. Is there a reason to not have a constructor like `ProductArray(P::ProductManifold, data) = ProductArray(ShapeSpecification(P.manifolds...), data)`?

Reply:
My first idea was to put the shape specification in the product manifold itself but since it is just for one point representation I thought it might be better to keep it separate. The main reason for not adding a constructor like that is because creating shape specification is relatively slow (I just benchmarked it and it takes about 8 microseconds for relatively simple cases).

Reply:
Yeah, I noticed it's slow. I guess the reason for this constructor would be convenience for the new user who's not worried about speed or only intends to create a single point for reuse. The documentation/examples could make it clear that this is not the efficient way to do it if one will want to create multiple points.

I agree it makes sense to keep the shape specification separate from the manifold. I didn't see it, but do we have a test where alternate representations are used? It's not clear to me how to do that.

Reply:
I think new users generally should use `ProductMPoint` instead of `ProductArray`, it's more intuitive.   I need to improve documentation on it. There are tests for `ProductMPoint` at the end of the test file for product manifolds.